<a href="https://colab.research.google.com/github/MariaArangoOchoa/Anal-tica-de-Negocios-/blob/main/Reto1_MariaArangoOchoa.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



**Modelo de PreAprobación - Naive Bayes**
0. Se procede con la carga de las librerías de trabajo

In [ ]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

1. Se cargan los datos

In [ ]:
nxl='/content/1. BD1_Longitudinal Survey.xlsx'
XDB=pd.read_excel(nxl,sheet_name=1)
XDB.dropna(inplace=True)


XDB=XDB[['Age', 'Mother_Edu', 'Father_Edu', 'FamilySize', 'Self_Esteem', 'Height', 'Weight', 'Income', 'Urban']]
XD=np.array(XDB[['Age', 'Mother_Edu', 'Father_Edu', 'FamilySize', 'Self_Esteem', 'Height', 'Weight', 'Income']])
yd=np.array(XDB['Urban'])

print(XDB)

tamano= len(XDB)

print("El tamaño de la base de datos es de: ",tamano)

XDB_urb = np.where(yd==1)[0]
len(XDB_urb)
pct_urb=len(XDB_urb)/tamano*100
print("El porcentaje de personas que viven el el area urbana es de: ", pct_urb, "%")

XDB_rur = np.where(yd==0)[0]
len(XDB_rur)
pct_rur=len(XDB_rur)/tamano*100
print("El porcentaje de personas que viven el el area rural es de: ", pct_rur, "%")

       Age  Mother_Edu  Father_Edu  FamilySize  Self_Esteem  Height  Weight  \
1       20         5.0         8.0           5         16.0    62.0   120.0   
5       19        12.0        12.0           4         27.0    64.0   200.0   
6       15        12.0        12.0           3         26.0    65.0   131.0   
7       21         9.0         6.0           3         23.0    65.0   179.0   
8       16        12.0        10.0           6         26.0    66.0   145.0   
...    ...         ...         ...         ...          ...     ...     ...   
12557   20        12.0        12.0           5         20.0    60.0   110.0   
12565   20        12.0        11.0           2         24.0    63.0   152.0   
12647   21        16.0        12.0           1         30.0    70.0   180.0   
12651   21        16.0        16.0           1         27.0    73.0   170.0   
12678   21         8.0         8.0           1         19.0    70.0   180.0   

        Income  Urban  
1          0.0    1.0  
5  

2. Implementamos Modelo

In [ ]:
mnb=GaussianNB()
mnb.fit(XD,yd)
ydp=mnb.predict(XD)

u=mnb.theta_
sigma=np.sqrt(mnb.var_)

LI=u-sigma
LS=u+sigma

print("la media de las variables por categoría:")
print(u)
print("El límite inferior de la variación de las variables por categoría es")
print(LI)
print("El límite superior de la variación de las variables por categoría es:")
print(LS)

cm=confusion_matrix(yd,ydp)
print("La matriz de confusión es:")
print(cm)


la media de las variables por categoría:
[[1.83810742e+01 1.07612958e+01 1.03324808e+01 4.74168798e+00
  2.20272805e+01 6.72148338e+01 1.48857630e+02 3.03672506e+04]
 [1.83569845e+01 1.11769401e+01 1.12565410e+01 4.68780488e+00
  2.26004435e+01 6.69031042e+01 1.44760976e+02 3.34278929e+04]]
El límite inferior de la variación de las variables por categoría es
[[ 1.58996876e+01  7.58892579e+00  6.46311334e+00  2.16621834e+00
   1.78344685e+01  6.31369799e+01  1.18746964e+02 -1.33465823e+03]
 [ 1.58811112e+01  7.79013864e+00  7.15097738e+00  2.24996381e+00
   1.84154526e+01  6.26897933e+01  1.14669654e+02 -2.69533531e+03]]
El límite superior de la variación de las variables por categoría es:
[[2.08624608e+01 1.39336659e+01 1.42018483e+01 7.31715762e+00
  2.62200925e+01 7.12926877e+01 1.78968296e+02 6.20691595e+04]
 [2.08328577e+01 1.45637416e+01 1.53621047e+01 7.12564594e+00
  2.67854344e+01 7.11164151e+01 1.74852297e+02 6.95511211e+04]]
La matriz de confusión es:
[[   2 1171]
 [   4 4506

3. Se procede con la evaluación del modelo


In [ ]:
VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]

Exactitud=(VP+VN)/len(XD)
print("La Exactitud es:", Exactitud)

TE=(FN+FP)/len(XD)
print("La Tasa de Error es:", TE)

Sensibilidad=VP/(VP+FP)
print("La Sensibilidad es:", Sensibilidad)

Especificidad=VN/(VN+FN)
print("La Especificidad es:", Especificidad)

Precisión=VP/(VP+FN)
print("La Precisión es:", Precisión)

PredNeg=VN/(VN+FP)
print("La Probabilidad de Negativo es:", PredNeg)

La Exactitud es: 0.7932430054548654
La Tasa de Error es: 0.2067569945451346
La Sensibilidad es: 0.793729082261758
La Especificidad es: 0.3333333333333333
La Precisión es: 0.9991130820399113
La Probabilidad de Negativo es: 0.0017050298380221654


4. Qué pasa con una sola persona

In [ ]:
Xun=np.array([[34,9,6,65,135,50000,30,3]])
ydun=mnb.predict(Xun)
print(ydun)

[1.]


**Análisis de Resultados:**

 De los resultados arrojados por el modelo, se puede observar que el número total de datos de la base de datos es 5683. Es pertinente mencionar que después de ver los resultados, el porcentaje de personas que viven en zonas urbanas es de 79.36%, a diferencia de un 20.64% de personas que viven en zona rural. Al ver los intervalos, fue interesante ver el comportamiento de dos variables, Edad y Tamaño Familiar que no cambiaron los límites inferiores y superiores entre las personas viviendo en zonas urbanas y rurales. Por otro lado, se puede analizar cómo la Educación de la Madre y la Altura fueron semejantes en sus resultados de urbano y rural, mas no iguales. Por el contrario, en peso e ingresos se ve una diferencia significativa. Más aún, se pudó encontrar valores muy altos respecto a la evaluación del modelo, con una precisión del 99.91%, una sensibilidad de 79.37% y una exactitud de 79.32%. No obstante, se vio una especificidad del 33.33%, con una tasa de error del 21.68%. Por último, la probabilidad de negativo fue de un 0.17%.

Adicionalmente, se pudo analizar el caso de una persona con unas características específicas y se concluyó por medio del modelo Naive Bayes que vivía en la ciudad.